File: structural/proxy.ts
// Sure! Let's use a real-world example of a virtual proxy for a heavy resource: a large image file.
// The proxy will load the image only when it is actually needed.

// In these examples, the ProxyImage class acts as a proxy to the RealImage class.
// The RealImage class represents a heavy resource that is expensive to create and load.
// The ProxyImage class controls access to the RealImage and ensures that
// it is only loaded when needed, optimizing resource usage.

// Subject interface
interface Image {
  display(): void;
}

// RealSubject
class RealImage implements Image {
  private filename: string;

  constructor(filename: string) {
    this.filename = filename;
    this.loadFromDisk();
  }

  display(): void {
    console.log(`Displaying ${this.filename}`);
  }

  private loadFromDisk(): void {
    console.log(`Loading ${this.filename} from disk...`);
  }
}

// Proxy
class ProxyImage implements Image {
  private filename: string;
  private realImage: RealImage | null = null;

  constructor(filename: string) {
    this.filename = filename;
  }

  display(): void {
    if (this.realImage === null) {
      this.realImage = new RealImage(this.filename);
    }
    this.realImage.display();
  }
}

// Client code
const image: Image = new ProxyImage("test_image.jpg");
image.display(); // Image will be loaded and displayed
image.display(); // Image will be displayed without loading

File: structural/proxy.py
from typing import Optional


# Subject interface
class Image:
    def display(self) -> None:
        pass


# RealSubject
class RealImage(Image):
    def __init__(self, filename: str) -> None:
        self.filename = filename
        self._load_from_disk()

    def display(self) -> None:
        print(f"Displaying {self.filename}")

    def _load_from_disk(self) -> None:
        print(f"Loading {self.filename} from disk...")


# Proxy
class ProxyImage(Image):
    def __init__(self, filename: str) -> None:
        self.filename = filename
        self._real_image: Optional[RealImage] = None

    def display(self) -> None:
        if self._real_image is None:
            self._real_image = RealImage(self.filename)
        self._real_image.display()


# Client code
image = ProxyImage("test_image.jpg")
image.display()  # Image will be loaded and displayed
image.display()  # Image will be displayed without loading
